src/aspire/source/__init__.py

```diff
@@ -370,6 +372,108 @@ def whiten(self, noise_filter):
         # Invalidate images
         self._im = None
 
+    def phase_flip(self):
+        """
+        Perform phase flip to images in the source object using CTF information
+        """
+        logger.info('Perform phase flip on source object')
+        logger.info('Adding Phase Flip Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(FlipXform(self.filters))
+        # Invalidate images
+        self._im = None
+
+    def reverse_contrast(self, batch_size=512):
+        """
+        Reverse the global contrast of images
+
+        Check if all images in a stack should be globally phase flipped so that
+        the molecule corresponds to brighter pixels and the background corresponds
+        to darker pixels. This is done by comparing the mean in a small circle
+        around the origin (supposed to correspond to the molecule) with the mean
+        of the noise, and making sure that the mean of the molecule is larger.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending a `Multiple` filter to the generation pipeline.
+
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info('Apply contrast reverse on source object')
+        L = self.L
+        image_center = (L + 1) / 2
+        coor_mat_m, coor_mat_n = np.meshgrid(np.arange(1, L + 1), np.arange(1, L + 1))
+
+        distance_from_center = np.sqrt((coor_mat_m - image_center) ** 2
+                                       + (coor_mat_n - image_center) ** 2)
+
+        # Get mask indices of signal and noise samples assuming molecule is around the center
+        signal_mask = distance_from_center < round(L / 4)
+        noise_mask = distance_from_center > round(L / 2 * 0.8)
+
+        # Calculate mean values in batch_size
+        signal_mean = 0.0
+        noise_mean = 0.0
+        for i in range(0, self.n, batch_size):
+            images = self.images(i, batch_size).asnumpy()
+            signal = (images * np.expand_dims(signal_mask, 2))
+            noise = (images * np.expand_dims(noise_mask, 2))
+            signal_denominator = self.n * np.sum(signal_mask)
+            noise_denominator = self.n * np.sum(noise_mask)
+            signal_mean += np.sum(signal) / signal_denominator
+            noise_mean += np.sum(noise) / noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to reverse contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to reverse contrast')
+            scale_factor = 1.0 * np.ones(self.n)
+
+        logger.info('Adding Scaling Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(Multiply(scale_factor))
+
+        # Invalidate images
+        self._im = None
+
+    def normalize_background(self, radius=None, batch_size=512):
```

Comment:
This works for now, but it might be good to have the option to do this on a per-image basis as well. Right now, this estimates the mean and standard deviation of the noise for the entire dataset, but these parameters may vary between images. Depending on the application, we may want to capture this or not. In addition, it may be relevant to consider a varying background mean (not just a constant) that we subtract from the images (sometimes known as ramp background subtraction).

Matlab code for doing this is found here: https://github.com/PrincetonUniversity/Cov3D/blob/b3afc69d2be262d9e3446fb89990c2b11664e0e5/matlab/imaging/im_normalize_bg.m.

Reply:
Done in the commit a92ddf2.

Reply:
The ramping option has be corrected and implemented in the commit 20d503f along with the option for normalization based on mean and std values from single image.